flink-runtime/src/test/java/org/apache/flink/runtime/leaderelection/ZooKeeperLeaderElectionConnectionHandlingTest.java

```diff
@@ -0,0 +1,162 @@
+
+        CompletableFuture<String> secondAddress = queueLeaderElectionListener.next();
+        assertThat("No result is expected since there was no leader elected before stopping the server, yet.", secondAddress, is(nullValue()));
+    }
+
+    @Test
+    public void testConnectionSuspendedHandling() throws Exception {
+        // initialize LeaderElection-related instances
+        String leaderAddress = "localhost";
+        LeaderElectionService leaderElectionService = ZooKeeperUtils.createLeaderElectionService(zooKeeperClient, config);
+        TestingContender contender = new TestingContender(leaderAddress, leaderElectionService);
+        leaderElectionService.start(contender);
+
+        // initialize LeaderRetrievalService
+        QueueLeaderElectionListener queueLeaderElectionListener = new QueueLeaderElectionListener(2, Duration.ofSeconds(1));
+
+        ZooKeeperLeaderRetrievalService testInstance = ZooKeeperUtils.createLeaderRetrievalService(zooKeeperClient, config);
+        testInstance.start(queueLeaderElectionListener);
+
+        // do the testing
+        CompletableFuture<String> firstAddress = queueLeaderElectionListener.next();
+        assertThat("The first result is expected to be the initially set leader address.", firstAddress.get(), is(leaderAddress));
+
+        stopTestServer();
+
+        CompletableFuture<String> secondAddress = queueLeaderElectionListener.next();
+        assertThat("The next result must not be missing.", secondAddress, not(is(nullValue())));
+        assertThat("The next result is expected to be null.", secondAddress.get(), is(nullValue()));
+    }
+
+    private void stopTestServer() throws IOException {
+        if (testingServer != null) {
+            testingServer.stop();
+            testingServer = null;
+        }
+    }
+
+    private static class QueueLeaderElectionListener implements LeaderRetrievalListener {
+
+        private final BlockingQueue<CompletableFuture<String>> queue;
+        private final Duration timeout;
+
+        public QueueLeaderElectionListener(int expectedCalls, Duration timeout) {
+            this.queue = new ArrayBlockingQueue<>(expectedCalls);
+            this.timeout = timeout;
+        }
+
+        @Override
+        public void notifyLeaderAddress(String leaderAddress, UUID leaderSessionID) {
+            try {
+                this.queue.offer(CompletableFuture.completedFuture(leaderAddress), timeout.toMillis(), TimeUnit.MILLISECONDS);
+            } catch (InterruptedException e) {
+                throw new IllegalStateException(e);
+            }
+        }
+
+        public CompletableFuture<String> next() {
+            try {
+                return this.queue.poll(timeout.toMillis(), TimeUnit.MILLISECONDS);
+            } catch (InterruptedException e) {
+                throw new IllegalStateException(e);
+            }
+        }
+
+        @Override
+        public void handleError(Exception exception) {
+            // nothing to do
```

Comment:
Let's throw the exception at least. Otherwise we might swallow it which could hide a problem and which makes debugging harder.